src/NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp

```diff
@@ -0,0 +1,195 @@
+#include "DataStructures/Transpose.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace {
+void apply_matrix_in_first_dim(double* result, const double* const input,
+                               const Matrix& matrix,
+                               const size_t size) noexcept {
+  dgemm_<true>('N', 'N',
+               matrix.rows(),            // rows of matrix and result
+               size / matrix.columns(),  // columns of result and input
+               matrix.columns(),         // columns of matrix and rows of input
+               1.0,                      // overall multiplier
+               matrix.data(),            // matrix
+               matrix.spacing(),         // rows of matrix including padding
+               input,                    // input
+               matrix.columns(),         // rows of input
+               0.0,                      // overwrite output with result
+               result,                   // result
+               matrix.rows());           // rows of result
+}
+}  // namespace
+
+template <typename SymmList, typename IndexList, size_t Dim>
+void logical_partial_derivative(
+    const gsl::not_null<TensorMetafunctions::prepend_spatial_index<
+        Tensor<DataVector, SymmList, IndexList>, Dim, UpLo::Lo,
+        Frame::Logical>*>
+        logical_derivative_of_u,
+    const gsl::not_null<gsl::span<double>*> buffer,
+    const Tensor<DataVector, SymmList, IndexList>& u,
+    const Mesh<Dim>& mesh) noexcept {
+  static_assert(
+      Dim > 0 and Dim < 4,
+      "logical_partial_derivative is only implemented for 1, 2, and 3d");
+  const size_t num_grid_points = mesh.number_of_grid_points();
+  ASSERT(buffer->size() >= num_grid_points,
+         "The buffer in logical_partial_derivative must be at least of size "
+             << num_grid_points << " but is of size " << buffer->size());
+
+  destructive_resize_components(logical_derivative_of_u,
+                                mesh.number_of_grid_points());
+  const Matrix identity{};
+  std::array<std::reference_wrapper<const Matrix>, Dim> diff_matrices{
+      make_array<Dim, std::reference_wrapper<const Matrix>>(identity)};
+  for (size_t d = 0; d < Dim; ++d) {
+    gsl::at(diff_matrices, d) =
+        std::cref(Spectral::differentiation_matrix(mesh.slice_through(d)));
+  }
+
+  for (size_t storage_index = 0; storage_index < u.size(); ++storage_index) {
+    const auto u_tensor_index = u.get_tensor_index(storage_index);
+    const auto xi_deriv_tensor_index = prepend(u_tensor_index, 0_st);
+    apply_matrix_in_first_dim(
+        logical_derivative_of_u->get(xi_deriv_tensor_index).data(),
+        u[storage_index].data(), diff_matrices[0], num_grid_points);
+    if constexpr (Dim > 1) {
```

Comment:
It seems like the dimensions after the first could be done in a loop without losing anything. The logic also looks pretty similar to what's done in `ApplyMatrices`,  and it would be nice to consolidate this kind of memory transpose and lapack call logic if at all possible.

Reply:
A bit more logic into computing the chunk size, but it does reduce number of lines :)